docs: break release notes into categories (#11217)
envoyproxy/envoy: docs/root/version_history/current.rst

1.15.0 (Pending)
================

> Incompatible Behavior Changes (things that expected to cause an incompatibility if they apply; changes required)

Reviewer: ```suggestion
Incompatible Behavior Changes (changes that are expected to cause an incompatibility if applicable; deployment changes are likely required)
```

Reviewer: style preference: I would probably put the description in italics under the header, same elsewhere.

-----------------------------

* build: official released binary is now built on Ubuntu 18.04, requires glibc >= 2.27.

> Minor Behavior Changes (things that may cause issues for some users, but should not for most)

Reviewer: ```suggestion
Minor Behavior Changes (changes that may cause incompatibilities for some users, but should not for most)
```

----------------

* access loggers: applied existing buffer limits to access logs, as well as :ref:`stats <config_access_log_stats>` for logged / dropped logs. This can be reverted temporarily by setting runtime feature `envoy.reloadable_features.disallow_unbounded_access_logs` to false.

> * http: fixed several bugs with applying correct connection close behavior across the http connection manager, health checker, and connection pool. This behavior may be temporarily reverted by setting runtime feature `envoy.reloadable_features.fix_connection_close` to false.

Reviewer 2: So do we want to dup issues like this to the bugfix section since they're both, or document (maybe in CONRIBUTING) that when changes overlap sections we add it to the first (as most important)?

Author: I think we'll have to iterate on this going forward. My current thought is let's put it into the first/most-important section.

* http: stopped overwriting `date` response headers. Responses without a `date` header will still have the header properly set. This behavior can be temporarily reverted by setting `envoy.reloadable_features.preserve_upstream_date` to false.
* http: fixed a bug where the upgrade header was not cleared on responses to non-upgrade requests.
  Can be reverted temporarily by setting runtime feature `envoy.reloadable_features.fix_upgrade_response` to false.
* http: stopped adding a synthetic path to CONNECT requests, meaning unconfigured CONNECT requests will now return 404 instead of 403. This behavior can be temporarily reverted by setting `envoy.reloadable_features.stop_faking_paths` to false.
* router: allow retries of streaming or incomplete requests. This removes stat `rq_retry_skipped_request_not_complete`.
* router: allow retries by default when upstream responds with :ref:`x-envoy-overloaded <config_http_filters_router_x-envoy-overloaded_set>`.

> Bug fixes (expected to improve things; unlikely to have negative effects)

Reviewer: ```suggestion
Bug fixes (changes expected to improve the state of the world and are unlikely to have negative effects)
```

---------

* http: fixed a bug in the grpc_http1_reverse_bridge filter where header-only requests were forwarded with a non-zero content length.
* http: fixed a bug where in some cases slash was moved from path to query string when :ref:`merging of adjacent slashes<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.merge_slashes>` is enabled.
* http: fixed several bugs with applying correct connection close behavior across the http connection manager, health checker, and connection pool. This behavior may be temporarily reverted by setting runtime feature `envoy.reloadable_features.fix_connection_close` to false.
* http: fixed a bug where the upgrade header was not cleared on responses to non-upgrade requests.
  Can be reverted temporarily by setting runtime feature `envoy.reloadable_features.fix_upgrade_response` to false.
* prometheus stats: fix the sort order of output lines to comply with the standard.
* upstream: fixed a bug where Envoy would panic when receiving a GRPC SERVICE_UNKNOWN status on the health check.

> Removed config or runtime

Reviewer: Perhaps link to the docs where we explain deprecations?

* router: allow retries by default when upstream responds with :ref:`x-envoy-overloaded <config_http_filters_router_x-envoy-overloaded_set>`.

Bug Fixes
---------
*Changes expected to improve the state of the world and are unlikely to have negative effects*

* http: fixed a bug in the grpc_http1_reverse_bridge filter where header-only requests were forwarded with a non-zero content length.
* http: fixed a bug where in some cases slash was moved from path to query string when :ref:`merging of adjacent slashes<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.merge_slashes>` is enabled.
* http: fixed several bugs with applying correct connection close behavior across the http connection manager, health checker, and connection pool. This behavior may be temporarily reverted by setting runtime feature `envoy.reloadable_features.fix_connection_close` to false.
* prometheus stats: fix the sort order of output lines to comply with the standard.
* upstream: fixed a bug where Envoy would panic when receiving a GRPC SERVICE_UNKNOWN status on the health check.

Removed Config or Runtime
-------------------------

> *Normally occurs at the end of the* :ref:`deprecation period <deprecated>`

Author: I couldn't figure out a way to have the `:ref:` be inside the italics/emphasis. If I put the closing `*` right after the closing backtick of the `:ref:` then it doesn't render as a link, but instead as the literal text `:ref: ....`. I also tried putting `**` around the link text inside the link, but that didn't work either. My final attempt was putting a space between the closing backtick and the closing `*`. That resulted in it not being emphasis rendered, but just having a literal `*` before and after the text.

Any other ideas are welcome, but for now I'm ok with this minor rendering bug in the docs.

Reviewer: np, thanks for trying.
